awx/main/models/credential/__init__.py

```diff
@@ -35,9 +40,10 @@
 from awx.main.utils import encrypt_field
 from . import injectors as builtin_injectors
 
-__all__ = ['Credential', 'CredentialType', 'V1Credential', 'build_safe_env']
+__all__ = ['Credential', 'CredentialType', 'CredentialInputSource', 'V1Credential', 'build_safe_env']
 
 logger = logging.getLogger('awx.main.models.credential')
+credential_plugins = [ep.load() for ep in iter_entry_points('awx.credential_plugins')]
```

Comment:
I should ask the obvious question of why `awx.credential_plugins`, as opposed to `awx.plugins.credentials`, although this might be more like a "credential input" plugin, because talking about credential plugins makes me think of credential types.

The answer is probably that `awx.plugins` is not an actual python module.

That gets to a larger issue about plugins, which I really do want to talk about. I know how notification template plugins are implemented, and it's not quite the direction that I would prefer to go in. Instead of looking at those as a model to start with building a new type of plugin system, I would look to:

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/loader.py

In particular, see `imp.load_source`. I think that the Ansible core team has a ton of things they can teach us about doing this right. They are really the experts, and have spent a huge amount of resources building out a plugin system that is flexible and works for everybody.

I think what their implementation shows us is that plugins really shouldn't be loaded by default as python modules. I know notifications use some metadata from the inherited class to make some decisions, and it might be harder to make those decisions on-demand.

Reply:
No particular reason - naming stuff is hard. `plugins.credentials` works for me.

Reply:
For what it's worth, this model of using `setuptools` entry points is built for _exactly_ this plugin model.  This approach is something that @matburt and I recommended to @jakemcdermott for this PR as lots of popular Python libraries use this as a mechanism for supporting and discovering third-party plugins.

Reply:
I think core's plugin model is grossly complex for our use case. The basic setuptools plugin entrypoints are the way to go. Notification Templates should have probably been implemented this way (and it's my fault they aren't)

Reply:
> awx.plugins.credentials

Nope. You're right. It is not a module. It is a shared namespace for plugins.

Reply:
I _do_ like `awx.plugins.credentials`.  Might as well solidify the namespace while we're here.

Reply:
gotcha, I see the reasoning for using it in the entry_points.